### PR TITLE
Fixes widget inspector can transform unmounted element

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

Our current widget inspector will try to visit ancestors of an error cause element in the case of a framework crash. The method that visits the ancestors of an element will throw an assertion error if the element is not mounted which results in another crash.

This pr exposes an api to get element's active state, and the widget inspector will use the api to determine if it should traverse the ancestors of the error causing element.

## Related Issues
https://github.com/flutter/flutter/issues/44485

## Tests

I added the following tests:

"test transformDebugCreator can transform unmounted element"

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -2660,6 +2660,11 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
   BuildOwner get owner => _owner;
   BuildOwner _owner;
 
+  /// Whether the element is active.
+  ///
+  /// A element will be active if it is mounted and attached to the element
+  /// tree.
+  bool get isActive => _active;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: We should not have this. Once an object is disposed, it should never be referenced, so there should be no need for it.

**Comment by Hixie** [comment]
I think the more fundamental problem here is that while we're in an unstable state, we try to walk the tree. This seems fundamentally like a dangerous operation. This particular failure is because we weren't active, but what if we were defunct? Or what if our parent pointer got corrupted? Or what if the tree ended up in a loop? There's all kinds of ways things could break, this is just one of many.

I would rather we just not walk the tree at all, or, if we are forced to for some reason, that we walk the tree in a way that is resilient to crashes.

**Comment by chunhtai** [comment]
> I think the more fundamental problem here is that while we're in an unstable state, we try to walk the tree. This seems fundamentally like a dangerous operation. This particular failure is because we weren't active, but what if we were defunct? Or what if our parent pointer got corrupted? Or what if the tree ended up in a loop? There's all kinds of ways things could break, this is just one of many.
> 
> I would rather we just not walk the tree at all, or, if we are forced to for some reason, that we walk the tree in a way that is resilient to crashes.

I think we can assume if element is active, it is in a safe state to walk the tree. Otherwise, I think we have a bigger problem.

But it sounds like we should build a more robust tree walking method that accounts for all the edge cases.

The current tree working logic visitAncestorElements is built in assumption every thing is stable and clean, otherwise throw assert. There is no api to know if the element or the tree is clean. I think this make sense because i feel this method is built in the mindset that it should only be called in some restricted case.

Maybe i should not have use this function at all?

**Comment by goderbauer** [comment]
Instead of walking the tree after a crash (where the tree is potentially unstable) could we collect the necessary information ("the relevant error-causing widget in user code") proactively while we build the tree? So when we crash, we can just have to do a simple look-up instead of walking the potentially broken tree?

**Comment by chunhtai** [comment]
> Instead of walking the tree after a crash (where the tree is potentially unstable) could we collect the necessary information ("the relevant error-causing widget in user code") proactively while we build the tree? So when we crash, we just have to do a simple look-up instead of walking the potentially broken tree?

This is doable. While we assign element.parent in mount we can also set a location field in element or its widget from parent.widget._location
 however we would probably need to make _Location and widget._location a public class, and I remember that is not something we want to expose, @jacob314 do you have any suggestion?

**Comment by jacob314** [comment]
I like 

> > Instead of walking the tree after a crash (where the tree is potentially unstable) could we collect the necessary information ("the relevant error-causing widget in user code") proactively while we build the tree? So when we crash, we just have to do a simple look-up instead of walking the potentially broken tree?
> 
> This is doable. While we assign element.parent in mount we can also set a location field in element or its widget from parent.widget._location
> however we would probably need to make _Location and widget._location a public class, and I remember that is not something we want to expose, @jacob314 do you have any suggestion?

The data should be cached on the Element not the Widget as the same widget could have multiple different ancestors in user code.

In principle this seems reasonable and could open up some other use cases only practical if the associated local Element is always available.

One challenge is that you might have to update the associated locally created element for an entire element  subtree when an element is mounted or unmounted which could cause performance problems.

Memory usage would be one extra field for each element object. That could be extra motivation to write the kernel transformer to eliminate debug only fields in release builds.

Making _Location public is fine. We could do that today without changing the kernel transformer by adding adding a base class that _Location extends that is public. The public API should expose just the location itself and not the parameter locations as it is not clear we want to continue supporting the parameter locations.
The _location field should not be made public but I'm fine with adding a top level method somewhere public in the package:flutter/widgets that exposes getting the Location object from Widget. Making the field public would be harmful as that would break anyone who defines a field in their Widget class named location.

**Comment by chunhtai** [comment]
That sounds reasonable. I will come up with a prototype soon

**Comment by Hixie** [comment]
If we do it proactively we'll have to be very mindful of performance. If we cache it we should be very mindful of correctness in cache invalidation.

Personally I would probably lean towards not doing it proactively but making the code that does the walk be resilient to badness and just give up silently if it can't walk the tree.

> I think we can assume if element is active, it is in a safe state to walk the tree. Otherwise, I think we have a bigger problem.

We do have a bigger problem, by definition. An exception just got thrown.

**Comment by jacob314** [comment]
Please keep me in the loop if you create some more resilient tree-walking helpers. They should also be used in the inspector where it would be nice to handle broken trees better.

**Comment by chunhtai** [comment]
@jacob314 do you have other use case the might make proactive user widget creation tracking useful?

**Comment by jacob314** [comment]
> @jacob314 do you have other use case the might make proactive user widget creation tracking useful?

Some half-baked ideas that could benefit from proactive tracking:
Visualizations that depend on knowing all the non-local widgets for each local widget.
* For example, imagine a debug-paint like visualization that connects up the local widgets to their implementation details. 
* Draw all bounding boxes for framework widgets associated with the selected local widget(s) in the inspector.

**Comment by chunhtai** [comment]
We should cache it for sure, otherwise, we have to walk the tree.
The cache can be updated without additional tree walk. There is only two point where we need to update the cache. First Element.mount when Element first created and Element.activate when Element got moved to a different location with global key. The global key reactively is done recursively, so we are walking the tree regardlessly. We just added an additional assignment.

We have to change dart kernel transform to also transform the Element base class if we want to keep all the Location class and in Widget_inspector.dart. I think it should be straight forward but I am not sure if there is any protocol i need to follow in order to do that.

**Comment by jacob314** [comment]
Changing the kernel transform for this seems like overkill. I'd recommend instead investing in a generic kernel transformer that removes debug only fields on release mode.

You also probably want a reference to the local Element not the local creation location if possible as that would be useful for more use cases.

**Comment by chunhtai** [comment]
> You also probably want a reference to the local Element not the local creation location if possible as that would be useful for more use cases.

I am not quite sure i understand what `a reference to the local Element not the local creation location` means.
The reason I can only think of change the transform kernel is that I will need to know about _Location and _HasUserCreatedWidgetLocation in framework.dart to add the caching logic. I can either import widget_inspector.dart to framework.dart or move the class declaration into framework.dart, but both seem weird to me.

**Comment by jacob314** [comment]
It means that in the non-local Element you would include a field that is the associated local Element ancestor. This of course assumes you are updating the data structures so that data is cleared out properly so you do not cause memory leaks.

Moving the class defined in the transformer is alright although you should be careful. `Location` is a common name so you would create a ton of name conflicts for users if you added that name as a public class name in package:flutter/widget.dart

**Comment by chunhtai** [comment]
I would have to update the kernal transformer if we want make the class public in framework.dart. The class name need to be change, I can update to something that is less likely to collide with user defined class

**Comment by goderbauer** [comment]
(PR triage): Closing this one for now since we will take a different approach.

**Pull request closed without merging**